### ISSUE TITLE
nrage trasfer pack code move in to core

### ISSUE THREAD

**Issue by project64** [project64/project64#189]
move the code out of nrage and in to the emulator core

**Comment by cxd4** [comment]
I didn't even know if that was possible since, iirc controller specs 1.0 have the transfer pak as raw-data-only?  So it would be handled only within the plugin but not the emulator.

But I have no knowledge of the MBC, transfer paks, or GameBoy hardware/games so never thought about it.  Just guessed it would be cleaner to Project64 source to keep that within the plugin to handle trivial accessories like that.  Pretty sure stuff like the microphone in _Hey, You, Pikachu!_ would also be better left to the plugin than the core since it's even more trivial support to contend with.

**Comment by project64** [comment]
I can always update the plugin spec :) 

if someone knew and had good code that they could legally contribute to the microphone for Hey, You, Pikachu I would be fine for that to be in the core as well.

**Comment by cxd4** [comment]
Well I just figured it'd be less work, less time off your hands to just not have to update the plugin spec just to be able to transfer more code out of some plugins into the core.  I'm not even sure if or why it would really matter whether an exact component is done only within the plugin, only within the core...I thought your concern with mempaks in N-Rage was that it was being handled by both when it should be one.  But which one to do it...only the core or the plugin, I don't see why it matters.  If anything the plugin would have more advantage to keep it because it could be tested to debug on other emulators.

**Comment by project64** [comment]
it means managing gameboy roms and eeproms can be controlled from the main settings window, tho it can use the settings from the emu if it wanted. Yes mempak needs to be merged for the duplicated reason, the moving the transfer code is just for nicety that the code is in one location.

**Comment by Frank-74** [comment]
waveInOpen function: https://msdn.microsoft.com/en-us/library/windows/desktop/dd743847%28v=vs.85%29.aspx

Might be useful for adding microphone input.

**Comment by cxd4** [comment]
First, if you had separate repositories for all the components (a repository for Project64, a repository for N-Rage, a repository for Glide64 or PJGlide64), this duplicated code problem wouldn't exactly be as much of a problem.  If mempak was handled by both N-Rage and also by PJGlide64, it would only matter to somebody whether they were contributing to the PJ64 repository or to the N-Rage repository.  Also people could fork/pull/do pull requests easier when the plugins have their own repository.

Second, if moving code out of the plugin and into the core is just for nicety, then wouldn't it be nice to have no code in the plugins and no plugins whatsoever and just do everything in the core not the plugins?  Why only the transfer pak...why even have a plugin system if it's just nice to move stuff like all of this outside of the plugin.

**Comment by weinerschnitzel** [comment]
I don't think this is such a bad thing. I agree that the N64 peripherals should be handled in the core. At first it makes sense to handle all of this in the input plugin, and it would make sense as far as making the core platform agnostic, but as far as other emulators go, removing the mempak code and handling it via the input plugin is a big change to plugin spec. 

Ideally I think an I/O plugin that handles saving/loading files and emulates controller input would be the best way to make the plugin spec platform agnostic, but I think that is another change that is too large. Might as well handle all file I/O in the core and leave the input plugin to handle just that.

**Comment by cxd4** [comment]
>removing the mempak code and handling it via the input plugin is a big change to plugin spec. 

It actually is no change at all to the plugin spec, since the spec has already been written to allow the mempak to be handled optionally by the core's personal implementation or within the plugin itself (raw data processing).  Within the plugin itself is more common for beyond-mempak processing since raw data is a more low level controller command interface to handle the SI functions.

My arguments of platform-agnosticism in Mupen64Plus changes to the plugin specs revolved around the inability to assume several facilities on all platforms, such as GUI forms and enumeration, which were never relevant to emulation itself.  In this case, mempak I/O is easily handled using the existence of a file system of any sort, which is guaranteed under the C standard library.  So I/O was less of an issue there, leaving no necessity to have a new plugin type just for I/O.

**Comment by weinerschnitzel** [comment]
Good point. I was thinking of changing the file paths, but that is completely different and doesn't apply.

**Comment by DaEckster** [comment]
Regarding Hey You! Pikachu, I've been discussing a possibility of doing an HLE version of it on the forums:
http://forum.pj64-emu.com/showthread.php?t=4951&page=2
http://www.emutalk.net/threads/55434-Calling-All-Adaptoid-Owners

However, right now I need someone with an Adaptoid to go through and find the responses from the VRU for each of the [459 possible in-game commands](http://pastebin.com/NaPmSBuR) ([640 with repeats](http://pastebin.com/ZRkdZQgy)).

**Comment by death-droid** [comment]
Made some progress on this in just porting over and cleaning up the code. Also been looking at other implementations of the MBC and the transferpak. 
Just wondering if anyone knows of any place that has some solid information on the transferpak itself, as it seems to be a poorly doccumented part of the N64

**Comment by Ecksters** [comment]
Information in the SDK on the Transfer Pak:
[http://level42.ca/projects/ultra64/Documentation/man/pro-man/pro26/26-07.html#04.1](http://level42.ca/projects/ultra64/Documentation/man/pro-man/pro26/26-07.html#04.1)

Also, I acquired an Adaptoid but discovered there appears to be no support built into the emulator(even with an Adaptoid) for the VRU. Had to set up a Windows XP machine just to test out the Adaptoid(no Vista+ drivers) and its software is specially designed for controllers only. I'd be willing to mail the VRU unit and mic to any developer who wants to seriously look into this.

I also contacted the [BlissBox](http://www.bliss-box.net/Bliss-Box/) developers and they said it's something they might look into in the future, but they'd need the hardware, so I might contact them again and see if they're willing, but I think we first need some way of passing raw controller data into the emulator in the controller settings, while still allowing other controllers to work(the Adaptoid plugin only assumes 1 controller, not a controller with the VRU plugged into the 4th port).
